docs/source/setup/environment.rst

```diff
@@ -86,12 +86,40 @@ Send Source
 Particularly during development, you may want to send files directly to workers
 that are already running.
 
-You should use ``client.upload_file`` in these cases.
+For single (or multiple) Python files, use ``client.upload_file``.
 For more detail, see the `API docs`_ and a
 StackOverflow question
 `"Can I use functions imported from .py files in Dask/Distributed?"`__
-This function supports both standalone file and setuptools's ``.egg`` files
-for larger modules.
+This function supports standalone files, as well as ``.zip`` archives and setuptools's ``.egg`` files
+for collections of Python files. For example, if you have a functions defined in a couple files
+``dask_funcs.py`` and ``utils.py``, you might do this in IPython or a Jupyter notebook::
+
+   !zip files.zip dask_funcs.py utils.py
+   client.upload_file("files.zip")
+
+   def test_import():
+       import dask_funcs
+       import utils
+       return "ok"
+   client.run(test_import)
+
+For an entire module, use ``client.pip_install`` with the ``local_packages`` argument (`docs <pipinstall_docs>`_).
+With this, you upload a source distribution or wheel of the local module to the cluster,
+then ``pip install`` it. Unlike ``client.upload_file``, this also installs all the dependencies
+specified in the ``setup.py`` file of your module.
+
+For example, if you're developing a package called ``mymod``, you might do this in IPython or a Jupyter notebook::
```

Comment:
Do you think it's worth warning folks about the problems with using `pip_install`. For example, worker restarts will not have the packages/files and same for adaptive scaling usage

Reply:
I'm pretty sure both worker restarts and new workers added to the cluster _will_ have the packages. When new workers join (including restarted ones, which are effectively new), the scheduler should tell them about all the existing worker plugins so they can catch up: https://github.com/dask/distributed/blob/4f83686b9ffed4e8c41bbf002c1bab24b38b9dab/distributed/scheduler.py#L4053

Reply:
Or are you implying that the resolve could be different Ben?

Reply:
No, i think I am mistaken about how the plugin worked.  I think this can be merged in.  Sorry for the noise